src/main/java/uk/gov/ida/verifylocalmatchingserviceexample/service/Cycle3MatchingService.java

```diff
@@ -0,0 +1,53 @@
+package uk.gov.ida.verifylocalmatchingserviceexample.service;
+
+import uk.gov.ida.verifylocalmatchingserviceexample.contracts.Cycle3AttributesDto;
+import uk.gov.ida.verifylocalmatchingserviceexample.contracts.MatchStatusResponseDto;
+import uk.gov.ida.verifylocalmatchingserviceexample.contracts.MatchingServiceRequestDto;
+import uk.gov.ida.verifylocalmatchingserviceexample.dataaccess.NationalInsuranceNumberDAO;
+import uk.gov.ida.verifylocalmatchingserviceexample.dataaccess.VerifiedPidDAO;
+import uk.gov.ida.verifylocalmatchingserviceexample.model.Person;
+
+import java.util.List;
+
+import static java.util.stream.Collectors.toList;
+import static uk.gov.ida.verifylocalmatchingserviceexample.contracts.MatchStatusResponseDto.MATCH;
+import static uk.gov.ida.verifylocalmatchingserviceexample.contracts.MatchStatusResponseDto.NO_MATCH;
+
+public class Cycle3MatchingService {
+
+    private static final String NATIONAL_INSURANCE_NUMBER = "nationalInsuranceNumber";
+    private NationalInsuranceNumberDAO nationalInsuranceNumberDAO;
+    private VerifiedPidDAO verifiedPidDAO;
+
+    public Cycle3MatchingService(NationalInsuranceNumberDAO nationalInsuranceNumberDAO, VerifiedPidDAO verifiedPidDAO) {
+        this.nationalInsuranceNumberDAO = nationalInsuranceNumberDAO;
+        this.verifiedPidDAO = verifiedPidDAO;
+    }
+
+    public MatchStatusResponseDto matchUser(MatchingServiceRequestDto matchingServiceRequest, List<Person> cycle1MatchingUsers) {
+        Cycle3AttributesDto cycle3AttributesDto = matchingServiceRequest.getCycle3AttributesDto();
+        if (!isNationalInsuranceNumberPresent(cycle3AttributesDto)) {
+            return NO_MATCH;
+        }
+
+        List<Integer> personIds = cycle1MatchingUsers.stream().map(Person::getPersonId).collect(toList());
+        List<Integer> matchingUserIds = nationalInsuranceNumberDAO
+            .getMatchingUserIds(personIds, cycle3AttributesDto.getAttributes().get(NATIONAL_INSURANCE_NUMBER));
+
+        if (matchingUserIds.size() == 1) {
+            verifiedPidDAO.save(matchingServiceRequest.getHashedPid(), matchingUserIds.get(0));
+            return MATCH;
```

Comment:
Maybe we could commonise the concept of "matching a user" to do the insertion and return MATCH at the same time? It is 2 lines though so probably not needed

Reply:
I wouldn't bother with this upon further reflection - it would probably add complexity given how boilerplate-y Java is.